internal/backend/runtime/omni/controllers/omni/machine_class_patch.go

```diff
@@ -0,0 +1,236 @@
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+
+package omni
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/controller/generic"
+	"github.com/cosi-project/runtime/pkg/resource"
+	"github.com/cosi-project/runtime/pkg/safe"
+	"github.com/cosi-project/runtime/pkg/state"
+	"github.com/siderolabs/gen/optional"
+	"github.com/siderolabs/gen/xslices"
+	"go.uber.org/zap"
+
+	"github.com/siderolabs/omni/client/pkg/cosi/labels"
+	"github.com/siderolabs/omni/client/pkg/omni/resources"
+	"github.com/siderolabs/omni/client/pkg/omni/resources/omni"
+	"github.com/siderolabs/omni/internal/backend/runtime/omni/controllers/helpers"
+)
+
+// MachineClassPatchControllerName is the name of the MachineClassPatchController.
+const MachineClassPatchControllerName = "MachineClassPatchController"
+
+// MachineClassPatchController builds mapping MachineSetNode -> MachineClass.
+type MachineClassPatchController struct {
+	generic.NamedController
+}
+
+// NewMachineClassPatchController initializes MachineClassPatchController.
+func NewMachineClassPatchController() *MachineClassPatchController {
+	return &MachineClassPatchController{
+		NamedController: generic.NamedController{
+			ControllerName: MachineClassPatchControllerName,
+		},
+	}
+}
+
+// Settings implements controller.QController interface.
+func (ctrl *MachineClassPatchController) Settings() controller.QSettings {
+	return controller.QSettings{
+		Inputs: []controller.Input{
+			{
+				Namespace: resources.DefaultNamespace,
+				Type:      omni.ConfigPatchType,
+				Kind:      controller.InputQPrimary,
+			},
+			{
+				Namespace: resources.DefaultNamespace,
+				Type:      omni.MachineClassType,
+				Kind:      controller.InputQMapped,
+			},
+			{
+				Namespace: resources.DefaultNamespace,
+				Type:      omni.MachineStatusType,
+				Kind:      controller.InputQMapped,
+			},
+		},
+		Outputs: []controller.Output{
+			{
+				Kind: controller.OutputShared,
+				Type: omni.ConfigPatchType,
+			},
+		},
+		Concurrency: optional.Some[uint](8),
+	}
+}
+
+// MapInput implements controller.QController interface.
+func (ctrl *MachineClassPatchController) MapInput(ctx context.Context, _ *zap.Logger,
+	r controller.QRuntime, ptr resource.Pointer,
+) ([]resource.Pointer, error) {
+	_, err := r.Get(ctx, ptr)
+	if err != nil {
+		if state.IsNotFoundError(err) {
+			return nil, nil
+		}
+	}
+
+	switch ptr.Type() {
+	case omni.MachineClassType:
+		patches, err := r.List(ctx, omni.NewConfigPatch(resources.DefaultNamespace, "").Metadata(),
+			state.WithLabelQuery(resource.LabelEqual(omni.LabelMachineClass, ptr.ID())))
+		if err != nil {
+			return nil, err
+		}
+
+		return xslices.Map(patches.Items, func(r resource.Resource) resource.Pointer {
+			return r.Metadata()
+		}), nil
+	case omni.MachineStatusType:
+		machineStatus, err := safe.ReaderGetByID[*omni.MachineStatus](ctx, r, ptr.ID())
```

Comment:
my only concern is that it's computationally intensive - how are we going to scale with that?

on any MachineStatus change we basically try to re-apply all machine class config patches

Reply:
Yeah, I tried to limit it only by the machines which are added to a cluster, but still.

Reply:
We can do an optimization I guess.
Can extract `MachineStatus` labels into a separate resource, this will lower the count of wake ups of the controller.
And this resource can also be used in the `MachineSetNode` controller that allocates the machines.